### PR TITLE
fix(web): properly handle selection 🎢

### PR DESCRIPTION
**Pull request by mcdurdin**
Relates to #5853 and others.

Selection management was not working properly with the various OutputTargets:

1. When there is a non-empty selection, rules have no context -- it's like new text.
2. Backspace over a selection deletes just the selection.
3. Typing a character replaces the selection, of course, and collapses the caret to the end of the new text.
4. `hasSelection` is a very strange name for `OutputTarget` descendants. It doesn't mean "has an active selection" but rather, kinda means "supports selection internally".
5. Added `isSelectionEmpty` which is used for some of the new selection rules above.

Note that the `touchAlias` OutputTarget class does not currently support selection. I hope we can deprecate `touchAlias` with the use of `inputMode` (#3030) in the future, rather than adding support for selection.

I tested selection and prediction click in 15.0.118 (when mouse interaction with suggestion banner was introduced) and it was not working correctly there -- it would delete text right of the selection unexpectedly. Prediction UI and selection will be fixed in an upcoming PR in this chain.

# User Testing

(Note: these tests are very good candidates for general acceptance tests!)

All these tests should be run in the KeymanWeb artifacts below.

* **GROUP_WEB_TEXTAREA:** Test in the TEXTAREA element in the KeymanWeb unminified test page.
* **GROUP_WEB_INPUT:** Test in the INPUT element in the KeymanWeb unminified test page.
* **GROUP_WEB_IFRAME:** Test in design-mode Iframes.

  To access design-mode iframes:
  1. Visit "Tests the new Attachment/Enablement API functionality" link in the KeymanWeb test artifacts.
  2. Click "Create design-mode IFrame" button.
  3. Click on the new "Edit me!" content shown

* **GROUP_WEB_EDITABLE_DIV:** Test in editable Divs.

  To access editable divs:
  1. Visit "Tests the new Attachment/Enablement API functionality" link in the KeymanWeb test artifacts.
  2. Click "Create editable DIV" button.
  3. Click on the new "Edit me!" content shown

**TEST_BASELINE:** Using a keyboard with some complex behaviours, such as Khmer Angkor, verify that input sequences are correct. For example, type <kbd>x</kbd><kbd>E</kbd><kbd>j</kbd><kbd>m</kbd><kbd>r</kbd> (desktop: <kbd>ខ</kbd><kbd>ែ</kbd><kbd>្</kbd><kbd>ម</kbd><kbd>រ</kbd> touch: <kbd>ខ</kbd><kbd>ែ (longpress េ)</kbd><kbd>្ម (longpress ម)</kbd><kbd>រ</kbd>) and verify that the emitted output is actually `ខ្មែរ` (you may need to use a tool to view the Unicode values, which should be as in the image below).

image

**TEST_SELECTION:** Using the web_context_tests keyboard, type <kbd>a</kbd><kbd>b</kbd><kbd>c</kbd><kbd>x</kbd>. Select `bc` and type <kbd>q</kbd>. The output after the final key should be `aqx`. (This is testing selection replacement with no matching keyboard rule.)

**TEST_SELECTION_2:** Using the web_context_tests keyboard, type <kbd>a</kbd><kbd>b</kbd><kbd>c</kbd><kbd>x</kbd>. Select `bc` and type <kbd>p</kbd>. The output after the final key should be `aqx`. (This is testing selection replacement with a matching keyboard rule.)

**TEST_CONTEXT_BASE:** Using the web_context_tests keyboard, type <kbd>a</kbd><kbd>b</kbd><kbd>c</kbd><kbd>d</kbd>. The output after the final key should be `!`.

**TEST_CONTEXT_SELECTION:** Using the attached web_context_tests keyboard, type <kbd>a</kbd><kbd>b</kbd><kbd>c</kbd><kbd>x</kbd>. Select the `x` character, and type <kbd>d</kbd>. The output after the final key should be `abcd`.

**TEST_CONTEXT_SELECTION_2:** Using the attached web_context_tests keyboard, type <kbd>a</kbd><kbd>b</kbd><kbd>c</kbd><kbd>x</kbd>. Select the `x` character, and delete it with <kbd>Backspace</kbd>. Type <kbd>d</kbd>. The output after the final key should be `!`.

**TEST_CONTEXT_SELECTION_3:** Using the attached web_context_tests keyboard, type <kbd>a</kbd><kbd>b</kbd><kbd>c</kbd><kbd>x</kbd>. Select the `x` character, and type <kbd>y</kbd>. Press <kbd>Backspace</kbd>, and type <kbd>d</kbd>. The output after the final key should be `!`.

**TEST_CONTEXT_SELECTION_4:** Using the attached web_context_tests keyboard, type <kbd>x</kbd><kbd>a</kbd><kbd>b</kbd><kbd>c</kbd><kbd>x</kbd>. Select the `abc` characters, and type <kbd>d</kbd>. The output after the final key should be `xdx`.

## SMP Tests

**TEST_SELECTION_SMP:** Using the web_context_tests keyboard, type <kbd>f</kbd><kbd>g</kbd><kbd>h</kbd><kbd>j</kbd>. Select `𐌁𐌂` and type <kbd>o</kbd>. The output after the final key should be `𐌀𐍈𐌈`.

**TEST_CONTEXT_BASE_SMP:** Using the web_context_tests keyboard, type <kbd>f</kbd><kbd>g</kbd><kbd>h</kbd><kbd>i</kbd>. The output after the final key should be `𐌙`.

**TEST_CONTEXT_SELECTION_SMP:** Using the attached web_context_tests keyboard, type <kbd>f</kbd><kbd>g</kbd><kbd>h</kbd><kbd>j</kbd>. Select the `𐌈` character, and type <kbd>i</kbd>. The output after the final key should be `𐌀𐌁𐌂i`.

**TEST_CONTEXT_SELECTION_SMP_2:** Using the attached web_context_tests keyboard, type <kbd>f</kbd><kbd>g</kbd><kbd>h</kbd><kbd>j</kbd>. Select the `𐌈` character, and delete it with <kbd>Backspace</kbd>. Type <kbd>i</kbd>. The output after the final key should be `𐌙`.

**TEST_CONTEXT_SELECTION_SMP_3:** Using the attached web_context_tests keyboard, type <kbd>f</kbd><kbd>g</kbd><kbd>h</kbd><kbd>j</kbd>. Select the `𐌈` character, and type <kbd>k</kbd>. Press <kbd>Backspace</kbd>, and type <kbd>i</kbd>. The output after the final key should be `𐌙`.

**TEST_CONTEXT_SELECTION_SMP_4:** Using the attached web_context_tests keyboard, type <kbd>k</kbd><kbd>f</kbd><kbd>g</kbd><kbd>h</kbd><kbd>k</kbd>. Select the `𐌀𐌁𐌂` characters, and type <kbd>i</kbd>. The output after the final key should be `𐌰i𐌰`.

### CODE REVIEW

File: web/source/dom/targets/input.ts [hunk 1]
```diff
@@ -55,13 +59,13 @@ namespace com.keyman.dom.targets {
...
-        return this.processedSelectionStart;
+        return this.processedSelectionEnd;
       } else {
         this._cachedSelectionStart = this.root.selectionStart; // KMW-1
         this.processedSelectionStart = this.root.value._kmwCodeUnitToCodePoint(this.root.selectionStart); // I3319
         this.processedSelectionEnd = this.root.value._kmwCodeUnitToCodePoint(this.root.selectionEnd);     // I3319
 
-        return this.processedSelectionStart;
+        return this.processedSelectionEnd;
       }
```
Comment (ermshiperete) [inline web/source/dom/targets/input.ts:69]: To me this code would be easier to understand like this:

```suggestion
      if (this.root.selectionStart != this._cachedSelectionStart) {
        this._cachedSelectionStart = this.root.selectionStart; // KMW-1
        this.processedSelectionStart = this.root.value._kmwCodeUnitToCodePoint(this.root.selectionStart); // I3319
        this.processedSelectionEnd = this.root.value._kmwCodeUnitToCodePoint(this.root.selectionEnd);     // I3319
      }
      return this.processedSelectionEnd;
```

File: web/source/dom/targets/textarea.ts [hunk 2]
```diff
@@ -64,13 +68,13 @@ namespace com.keyman.dom.targets {
...
-        return this.processedSelectionStart;
+        return this.processedSelectionEnd;
       } else {
         this._cachedSelectionStart = this.root.selectionStart; // KMW-1
         this.processedSelectionStart = this.root.value._kmwCodeUnitToCodePoint(this.root.selectionStart); // I3319
         this.processedSelectionEnd = this.root.value._kmwCodeUnitToCodePoint(this.root.selectionEnd);     // I3319
 
-        return this.processedSelectionStart;
+        return this.processedSelectionEnd;
       }
```
Comment (ermshiperete) [inline web/source/dom/targets/textarea.ts:78]: To me this code would be easier to understand like this:

```suggestion
      if (this.root.selectionStart != this._cachedSelectionStart) {
        this._cachedSelectionStart = this.root.selectionStart; // KMW-1
        this.processedSelectionStart = this.root.value._kmwCodeUnitToCodePoint(this.root.selectionStart); // I3319
        this.processedSelectionEnd = this.root.value._kmwCodeUnitToCodePoint(this.root.selectionEnd);     // I3319
      }
      return this.processedSelectionEnd;
```

**Review by ermshiperete — APPROVED** [review]
lgtm as far as I can tell :smile:

File: web/unit_tests/CI.conf.js [hunk 3]
```diff
@@ -92,12 +92,6 @@ module.exports = function(config) {
       os_version: '10',
       browser: 'chrome',
       browser_version: '70.0'
-    },
-    bs_edge_win: {
-      os: 'Windows',
-      os_version: '10',
-      browser: 'edge',
-      version: '18.0'
```
Comment (mcdurdin) [inline web/unit_tests/CI.conf.js:100]: We no longer support legacy Edge, and it fails some selection tests due to bugs in its selectionDirection support

File: web/unit_tests/cases/element_interfaces.js [hunk 4]
```diff
@@ -423,7 +423,7 @@ if(typeof InterfaceTests == 'undefined') {
       String.kmwEnableSupplementaryPlane(false);
       testObj.resetWithText(pair, Apple.normal);
       testObj.setSelectionRange(pair, 3, 5);
-      assert.equal(pair.wrapper.getCaret(), 3, "Failed to choose the caret's initial position for forward-selections within a simple string");
+      assert.equal(pair.wrapper.getCaret(), 5, "Failed to choose the caret's initial position for forward-selections within a simple string");
```
Comment (mcdurdin) [inline web/unit_tests/cases/element_interfaces.js:426]: Caret position is at end of selection for forward selection.

File: web/unit_tests/cases/element_interfaces.js [hunk 5]
```diff
@@ -437,7 +437,7 @@ if(typeof InterfaceTests == 'undefined') {
       String.kmwEnableSupplementaryPlane(true);
       testObj.resetWithText(pair, Apple.smp);
       testObj.setSelectionRange(pair, 6, 10);
-      assert.equal(pair.wrapper.getCaret(), 3, "Failed to caret's initial position for forward-selections within an SMP string");
+      assert.equal(pair.wrapper.getCaret(), 5, "Failed to caret's initial position for forward-selections within an SMP string");
```
Comment (mcdurdin) [inline web/unit_tests/cases/element_interfaces.js:440]: Caret position is at end of selection for forward selection.

File: web/source/dom/targets/input.ts [hunk 6]
```diff
@@ -54,53 +58,58 @@ namespace com.keyman.dom.targets {
...
+    }
+
+    getSelectionDirection(): "forward" | "backward" | "none" {
+      return this.root.selectionDirection;
     }
 
     getTextBeforeCaret(): string {
-      return this.getText()._kmwSubstring(0, this.getCaret());
+      this.getCaret();
+      return this.getText()._kmwSubstring(0, this.processedSelectionStart);
```
Comment (ermshiperete) [inline web/source/dom/targets/input.ts:94]: Does this have to take the `selectionDirection` into consideration?

File: web/source/dom/targets/input.ts [hunk 7]
```diff
@@ -54,53 +58,58 @@ namespace com.keyman.dom.targets {
...
     }
 
     setTextBeforeCaret(text: string) {
       this.getCaret();
+      let selectionLength = this.processedSelectionEnd - this.processedSelectionStart;
```
Comment (ermshiperete) [inline web/source/dom/targets/input.ts:99]: Does this have to take `selectionDirection` into consideration so that we don't end up with a negative length?

**Review by ermshiperete — COMMENTED** [review]
Oh the joys of selection directions :smile: 

I wonder if it would make things easier to understand and implement if we'd introduce two additional properties, e.g. `selectionTop` and `selectionBottom`, that point to the first and last character/position of a selection, counted from the beginning of the text. `selectionTop` would return `selectionStart` or `selectionEnd`, depending on the direction of the selection.

**Comment by mcdurdin** [comment]
> I wonder if it would make things easier to understand and implement if we'd introduce two additional properties, e.g. `selectionTop` and `selectionBottom`, that point to the first and last character/position of a selection, counted from the beginning of the text. `selectionTop` would return `selectionStart` or `selectionEnd`, depending on the direction of the selection.

Fortunately, we don't have to worry about this. `selectionStart <= selectionEnd` is always true. `selectionDirection` determines which end of the selection has the caret (`forward` means the caret is at `selectionEnd`, and `backward` means the caret is at `selectionStart`; `none` means there is no caret, e.g. on iOS when selecting with finger, or no selection, just a caret, on many devices).

File: web/source/dom/targets/input.ts (hunk 7 above)
Comment (mcdurdin) [reply web/source/dom/targets/input.ts:99]: No -- see my previous comment.

File: web/source/dom/targets/input.ts (hunk 6 above)
Comment (mcdurdin) [reply web/source/dom/targets/input.ts:94]: No -- see my previous comment.

**Comment by MakaraSok** [comment]
# User Test Results

* **GROUP_WEB_TEXTAREA: Test in the TEXTAREA element in the KeymanWeb unminified test page.**
All tests have been done at [unminified.html](https://build.palaso.org/repository/download/Keymanweb_TestPullRequests/308277:id/testing/unminified.html) using Safari 15.3 on macOS 12.2.
text area<br/>
  * **TEST_BASELINE (PASSED)** The output is indeed `ខ្មែរ`.
  * **TEST_SELECTION (PASSED)** The behavior is shown as indicated, but the OSK is empty. The output of each key is like that of a QWERTY keyboard layout. The same empty OSK is observed when testing on Chrome in macOS.
  * **TEST_CONTEXT_BASE (PASSED)** The behavior is shown as indicated, but the OSK is empty. The output of each key is like that of a QWERTY keyboard layout.
  * **TEST_CONTEXT_SELECTION (PASSED)** The behavior is shown as indicated, but again the OSK stays empty. The output of each key is like that of a QWERTY keyboard layout.
  * **TEST_CONTEXT_SELECTION_2 (PASSED)** The behavior is shown as indicated, but again the OSK stays empty. The output of each key is like that of a QWERTY keyboard layout.
  * **TEST_CONTEXT_SELECTION_3 (PASSED)** The behavior is shown as indicated, but again the OSK stays empty. The output of each key is like that of a QWERTY keyboard layout.
  * **TEST_CONTEXT_SELECTION_4 (PASSED)** The behavior is shown as indicated, but again the OSK stays empty. The output of each key is like that of a QWERTY keyboard layout.

* **GROUP_WEB_INPUT: Test in the INPUT element in the KeymanWeb unminified test page.**
All tests have been done at [unminified.html](https://build.palaso.org/repository/download/Keymanweb_TestPullRequests/308277:id/testing/unminified.html) using Safari 15.3 on macOS 12.2.
input field<br/>
  * **TEST_BASELINE (PASSED)** The output is indeed `ខ្មែរ`.
  * **TEST_SELECTION (PASSED)** The behavior is shown as indicated, but the OSK is empty. The output of each key is like that of a QWERTY keyboard layout. The same empty OSK is observed when testing on Chrome in macOS.
  * **TEST_CONTEXT_BASE (PASSED)** The behavior is shown as indicated, but the OSK is empty. The output of each key is like that of a QWERTY keyboard layout.
  * **TEST_CONTEXT_SELECTION (PASSED)** The behavior is shown as indicated, but again the OSK stays empty. The output of each key is like that of a QWERTY keyboard layout.
  * **TEST_CONTEXT_SELECTION_2 (PASSED)** The behavior is shown as indicated, but again the OSK stays empty. The output of each key is like that of a QWERTY keyboard layout.
  * **TEST_CONTEXT_SELECTION_3 (PASSED)** The behavior is shown as indicated, but again the OSK stays empty. The output of each key is like that of a QWERTY keyboard layout.
  * **TEST_CONTEXT_SELECTION_4 (PASSED)** The behavior is shown as indicated, but again the OSK stays empty. The output of each key is like that of a QWERTY keyboard layout.

* **GROUP_WEB_IFRAME: Test in design-mode Iframes.**
All tests have been done at [unminified.html](https://build.palaso.org/repository/download/Keymanweb_TestPullRequests/308277:id/testing/attachment-api/?mode=auto) using Safari 15.3 on macOS 12.2.
Design-mode IFrames<br/>
  * **TEST_BASELINE (PASSED)** The output is indeed `ខ្មែរ`.
  * **TEST_SELECTION (PASSED)** The behavior is shown as indicated, but the OSK is empty. The output of each key is like that of a QWERTY keyboard layout. The same empty OSK is observed when testing on Chrome in macOS.
  * **TEST_CONTEXT_BASE (PASSED)** The behavior is shown as indicated, but the OSK is empty. The output of each key is like that of a QWERTY keyboard layout.
  * **TEST_CONTEXT_SELECTION (PASSED)** The behavior is shown as indicated, but again the OSK stays empty. The output of each key is like that of a QWERTY keyboard layout.
  * **TEST_CONTEXT_SELECTION_2 (PASSED)** The behavior is shown as indicated, but again the OSK stays empty. The output of each key is like that of a QWERTY keyboard layout.
  * **TEST_CONTEXT_SELECTION_3 (PASSED)** The behavior is shown as indicated, but again the OSK stays empty. The output of each key is like that of a QWERTY keyboard layout.
  * **TEST_CONTEXT_SELECTION_4 (PASSED)** The behavior is shown as indicated, but again the OSK stays empty. The output of each key is like that of a QWERTY keyboard layout.

* **GROUP_WEB_EDITABLE_DIV: Test in editable Divs.**
All tests have been done at [unminified.html](https://build.palaso.org/repository/download/Keymanweb_TestPullRequests/308277:id/testing/attachment-api/?mode=auto) using Safari 15.3 on macOS 12.2.
contenteditable Elements<br/>
  * **TEST_BASELINE (PASSED)** The output is indeed `ខ្មែរ`.
  * **TEST_SELECTION (PASSED)** The behavior is shown as indicated, but the OSK is empty. The output of each key is like that of a QWERTY keyboard layout. The same empty OSK is observed when testing on Chrome in macOS.
  * **TEST_CONTEXT_BASE (PASSED)** The behavior is shown as indicated, but the OSK is empty. The output of each key is like that of a QWERTY keyboard layout.
  * **TEST_CONTEXT_SELECTION (PASSED)** The behavior is shown as indicated, but again the OSK stays empty. The output of each key is like that of a QWERTY keyboard layout.
  * **TEST_CONTEXT_SELECTION_2 (PASSED)** The behavior is shown as indicated, but again the OSK stays empty. The output of each key is like that of a QWERTY keyboard layout.
  * **TEST_CONTEXT_SELECTION_3 (PASSED)** The behavior is shown as indicated, but again the OSK stays empty. The output of each key is like that of a QWERTY keyboard layout.
  * **TEST_CONTEXT_SELECTION_4 (PASSED)** The behavior is shown as indicated, but again the OSK stays empty. The output of each key is like that of a QWERTY keyboard layout.

File: web/unit_tests/cases/text_selection.js [hunk 8]
```diff
@@ -0,0 +1,331 @@
...
+          await assertInputSteps(ele, 4, () => {
+            eventDriver.simulateEvent(keys.A);
+            eventDriver.simulateEvent(keys.B);
+            eventDriver.simulateEvent(keys.C);
+            eventDriver.simulateEvent(keys.X);
+          });
+
+          assert.strictEqual(ele.value, 'abcx', "Sanity check: expected text to be 'abcx'");
+
+          // Step 2, select 'bc', replace with 'q'
```
Comment (ermshiperete) [inline web/unit_tests/cases/text_selection.js]:
```suggestion
          // Step 2, select 'x', replace with 'd'
```

File: web/unit_tests/cases/text_selection.js [hunk 9]
```diff
@@ -0,0 +1,331 @@
...
+          assert.deepEqual([ele.selectionStart, ele.selectionEnd], [2, 2], "Expected caret to be after 'q'");
+        });
+
+        /**
+         * TEST_CONTEXT_SELECTION
+         *
+         * Using the attached web_context_tests keyboard, type abcx. Select the x character, and type d.
+         * The output after the final key should be 'abcd'.
+         */
+        it('Should ignore context when a selection is made', async () => {
```
Comment (ermshiperete) [inline web/unit_tests/cases/text_selection.js:227]:
```suggestion
        it('Should ignore context when a '+direction+' selection is made', async () => {
```

File: web/unit_tests/cases/text_selection.js [hunk 10]
```diff
@@ -0,0 +1,331 @@
...
+          assert.deepEqual([ele.selectionStart, ele.selectionEnd], [4, 4], "Expected caret to be after 'd'");
+        });
+
+        /**
+         * TEST_CONTEXT_SELECTION_2
+         *
+         * Using the attached web_context_tests keyboard, type abcx. Select the x character, and
+         * delete it with Backspace. Type d. The output after the final key should be !.
+         */
+        it('Should correctly delete selection with backspace and not lose sync', async () => {
```
Comment (ermshiperete) [inline web/unit_tests/cases/text_selection.js:293]:
```suggestion
        it('Should correctly delete '+direction+' selection with backspace and not lose sync', async () => {
```

File: web/unit_tests/cases/text_selection.js [hunk 11]
```diff
@@ -0,0 +1,331 @@
...
+          assert.deepEqual([ele.selectionStart, ele.selectionEnd], [1, 1], "Expected caret to be after '!'");
+        });
+
+        /**
+         * TEST_CONTEXT_SELECTION_3
+         *
+         * Using the attached web_context_tests keyboard, type abcx. Select the x character, and type y.
+         * Press Backspace, and type d. The output after the final key should be !.
+         */
+        it('Should correctly repplace selection and not lose sync', async () => {
```
Comment (ermshiperete) [inline web/unit_tests/cases/text_selection.js]:
```suggestion
        it('Should correctly replace '+direction+' selection and not lose sync', async () => {
```

File: web/unit_tests/cases/text_selection.js [hunk 12]
```diff
@@ -0,0 +1,331 @@
...
+          assert.deepEqual([ele.selectionStart, ele.selectionEnd], [1, 1], "Expected caret to be after '!'");
+        });
+
+        /**
+         * TEST_CONTEXT_SELECTION_4
+         *
+         * Using the attached web_context_tests keyboard, type xabcx.
+         * Select the abc characters, and type d. The output after the final key should be xdx.
+         */
+        it('Should not treat the selection as context', async () => {
```
Comment (ermshiperete) [inline web/unit_tests/cases/text_selection.js:481]:
```suggestion
        it('Should not treat the '+direction+' selection as context', async () => {
```

**Review by ermshiperete — APPROVED** [review]

**Comment by mcdurdin** [comment]
I ran these tests against earlier versions, just to see what the how editing was going on earlier versions. The selection issues are certainly much worse on 15.0-alpha on Android and iOS, but we don't have the capability of auto-test on those platforms at this point.

## 14.0.286-stable

    √ 28 tests completed
    × 16 tests failed

    FAILED TESTS:
      × Should ignore context when a selection is made
      × Should ignore context when a selection is made - SMP
      × Should correctly delete selection with backspace and not lose sync
      × Should correctly delete selection with backspace and not lose sync - SMP
      × Should ignore context when a selection is made
      × Should ignore context when a selection is made - SMP
      × Should correctly delete selection with backspace and not lose sync
      × Should correctly delete selection with backspace and not lose sync - SMP

## 15.0.120-alpha through 15.0.202-alpha

    √ 28 tests completed
    × 16 tests failed

    FAILED TESTS:
      × Should ignore context when a selection is made
      × Should ignore context when a selection is made - SMP
      × Should correctly delete selection with backspace and not lose sync
      × Should correctly delete selection with backspace and not lose sync - SMP
      × Should ignore context when a selection is made
      × Should ignore context when a selection is made - SMP
      × Should correctly delete selection with backspace and not lose sync
      × Should correctly delete selection with backspace and not lose sync - SMP

**Comment by bharanidharanj** [comment]
* **GROUP_WEB_IFRAME**: **Test in design-mode Iframes**

   * **TEST_SELECTION_SMP (PASSED):** The output is showing 𐌀𐍈𐌈.
   * **TEST_CONTEXT_BASE_SMP (PASSED):** The output is showing 𐌙 after typing fghi
   * **TEST_CONTEXT_SELECTION_SMP (PASSED):** The output is showing 𐌀𐌁𐌂i after typing i
   * **TEST_CONTEXT_SELECTION_SMP2 (PASSED):** The output is showing 𐌙 after typing i
   * **TEST_CONTEXT_SELECTION_SMP3 (PASSED):** The output is showing 𐌙 after typing i
   * **TEST_CONTEXT_SELECTION_SMP4 (PASSED):** The output is showing 𐌰i𐌰 after typing i

**Comment by bharanidharanj** [comment]
* **GROUP_WEB_EDITABLE_DIV**: **Test in editable Divs**

   * **TEST_SELECTION_SMP (PASSED):** The output is showing 𐌀𐍈𐌈.
   * **TEST_CONTEXT_BASE_SMP (PASSED):** The output is showing 𐌙 after typing fghi 
   * **TEST_CONTEXT_SELECTION_SMP (PASSED):** The output is showing 𐌀𐌁𐌂i after typing i.
   * **TEST_CONTEXT_SELECTION_SMP_2 (PASSED):** The output is showing 𐌙 after typing i
   * **TEST_CONTEXT_SELECTION_SMP_3 (PASSED):** The output is showing 𐌙 after typing i.
   * **TEST_CONTEXT_SELECTION_SMP_4 (PASSED):** The output is showing 𐌰i𐌰 after typing i.

**Comment by bharanidharanj** [comment]
* **GROUP_WEB_TEXTAREA**: **Test in the TEXTAREA element in the KeymanWeb unminified test page**

   * **TEST_SELECTION_SMP (PASSED):** The output is showing 𐌀𐍈𐌈.
   * **TEST_CONTEXT_BASE_SMP (PASSED):** The output is showing 𐌙 after typing fghi 
   * **TEST_CONTEXT_SELECTION_SMP (PASSED):** The output is showing 𐌀𐌁𐌂i after typing i.
   * **TEST_CONTEXT_SELECTION_SMP_2 (PASSED):** The output is showing 𐌙 after typing i
   * **TEST_CONTEXT_SELECTION_SMP_3 (PASSED):** The output is showing 𐌙 after typing i.
   * **TEST_CONTEXT_SELECTION_SMP_4 (PASSED):** The output is showing 𐌰i𐌰 after typing i.

**Comment by bharanidharanj** [comment]
* **GROUP_WEB_INPUT**: **Test in the INPUT element in the KeymanWeb unminified test page**

   * **TEST_SELECTION_SMP (PASSED):** The output is showing 𐌀𐍈𐌈.
   * **TEST_CONTEXT_BASE_SMP (PASSED):** The output is showing 𐌙 after typing fghi 
   * **TEST_CONTEXT_SELECTION_SMP (PASSED):** The output is showing 𐌀𐌁𐌂i after typing i.
   * **TEST_CONTEXT_SELECTION_SMP_2 (PASSED):** The output is showing 𐌙 after typing i
   * **TEST_CONTEXT_SELECTION_SMP_3 (PASSED):** The output is showing 𐌙 after typing i.
   * **TEST_CONTEXT_SELECTION_SMP_4 (PASSED):** The output is showing 𐌰i𐌰 after typing i.

**Comment by bharanidharanj** [comment]
* **GROUP_WEB_TEXTAREA**: **Test in the TEXTAREA element in the KeymanWeb unminified test page**

   * **TEST_SELECTION_2 (PASSED):** The output is showing aqx after typing p.


* **GROUP_WEB_INPUT**: **Test in the INPUT element in the KeymanWeb unminified test page**

   * **TEST_SELECTION_2 (PASSED):** The output is showing aqx after typing p.

**Comment by bharanidharanj** [comment]
* **GROUP_WEB_IFRAME**: **Test in design-mode Iframes.**

   * **TEST_SELECTION_2 (PASSED):** The output is showing aqx after typing p.


* **GROUP_WEB_EDITABLE_DIV**: **Test in editable Divs.
   * **TEST_SELECTION_2 (PASSED):** The output is showing aqx after typing p.

**Pull request merged**